text/0031-integ-testing.md

```diff
@@ -0,0 +1,300 @@
+# Integration testing for CDK apps
+
+* **Original Author(s):**: [@nija-at](https://github.com/nija-at)
+* **Tracking Issue**: [#31](https://github.com/aws/aws-cdk-rfcs/issues/31)
+* **API Bar Raiser**: @{BAR_RAISER_USER}
+
+CDK app developers and construct library authors can write integration
+tests that can be run locally or in their CI/CD pipeline.
+
+## Working Backwards - README.md
+
+### contest
+
+🌩️ contest tests can verify that your CDK app can be deployed successfully on AWS with the desired resources.
+
+🤖 contest tests can verify that your CDK app functions as expected, such as assertions against a public API endpoint.
+
+🕰️ contest tests can verify that your CDK app can be updated from a previous version of the app.
+
+🧑‍💻 Write contest tests in any CDK supported language, and execute them as part of your CI pipeline.
+
+📸 Capture snapshots of your CDK apps to track how they are impacted by changes.
+
+### Writing Tests
+
+contest tests can be written in the same programming language as your CDK project.
+
+Let's start with the below CDK construct that creates an artifact repository in an S3 bucket,
+and notifies of new artifacts to an SQS queue.
+
+```ts
+// repo.ts
+import { Construct } from 'constructs';
+import { Bucket, EventType, IBucket } from '@aws-cdk/aws-s3';
+import { IQueue, Queue } from '@aws-cdk/aws-sqs';
+import { SqsDestination } from '@aws-cdk/aws-s3-notifications';
+
+export class ArtifactRepo extends Construct {
+  public readonly artifacts: IBucket;
+  public readonly publishNotifs: IQueue;
+  
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.artifacts = new Bucket(this, 'Artifacts');
+    this.publishNotifs = new Queue(this, 'PublishNotifs');
+
+    this.artifacts.addEventNotification(EventType.OBJECT_CREATED_PUT,
+      new SqsDestination(this.publishNotifs))
+  }
+}
+```
+
+The following integ test invokes puts an object in the S3 bucket and verifies the message in the SQS queue.
+
+```ts
+// repo.contest.ts
+import { ArtifactRepo } from './records';
+import { Test } from '@aws-cdk/contest';
+
+const repo = new ArtifactRepo(app, 'RequestRecord');
+
+const repoTest = Test.forConstruct(repo);
+
+repoTest.assertAws.
+  s3.putObject({ bucket: stack.artifacts, key: '...', body: '...' });
+
+repoTest.assertAws.
+  sqs.receiveMessage({ queue: stack.publishNotifs })
+     .returns({ Messages: [ ... ] });
+```
+
+### Assertions
+
+`contest` comes canned with a set of commonly used high level assertions.
+
+* `assertAws`: execute AWS service APIs and assert the response.
+* `invokeLambda`: invoke an AWS lambda function and assert the response.
+* `curlApi`: Execute the 'curl' command on an API endpoint. Used typically to test API Gateway resources.
+
+By default, these assertions will check that the underlying call succeeds. That behavior can be extended
+to also include verification of the response.
+
+```ts
+test.assertAws(
```

Comment:
Ideally we want a programming model that will encourage an open ecosystem of assertions up to the point that even `assertAws` is not part of the AWS CDK but vended as a separate library. This likely implies that the APIs have separate entrypoints:

```ts
import { AwsAssertions } from 'cdk-aws-assertions';

const assert new AwsAssertions(test);
assert.s3.getObject({ ... }).returns(...);
```

Something like that.